dowhy/causal_prediction/algorithms/cacm.py

```diff
@@ -0,0 +1,169 @@
+        :param attr_types: list of attribute types (based on relationship with label Y); should be ordered according to attribute order in loaded dataset.
+            Currently, 'causal' (Causal), 'conf' (Confounded), 'ind' (Independent) and 'sel' (Selected) are supported.
+            For single-shift datasets, use: ['causal'], ['ind']
+            For multi-shift datasets, use: ['causal', 'ind']
+        :param E_conditioned: Binary flag indicating whether E-conditioned regularization has to be applied
+        :param E_eq_A: list indicating indices of attributes that coincide with environment (E) definition; default is empty.
+        :param gamma: kernel bandwidth for MMD (due to implementation, the kernel bandwdith will actually be the reciprocal of gamma i.e., gamma=1e-6 implies kernel bandwidth=1e6. See `mmd_compute` in utils.py)
+        :param lambda_causal: MMD penalty hyperparameter for Causal shift
+        :param lambda_conf: MMD penalty hyperparameter for Confounded shift
+        :param lambda_ind: MMD penalty hyperparameter for Independent shift
+        :param lambda_sel: MMD penalty hyperparameter for Selected shift
+        :returns: an instance of PredictionAlgorithm class
+
+        """
+
+        super().__init__(model, optimizer, lr, weight_decay, betas, momentum)
+
+        self.CACMRegularization = Regularization(E_conditioned, ci_test, kernel_type, gamma)
+
+        self.attr_types = attr_types
+        self.E_eq_A = E_eq_A
+        self.lambda_causal = lambda_causal
+        self.lambda_conf = lambda_conf
+        self.lambda_ind = lambda_ind
+        self.lambda_sel = lambda_sel
+
+    def training_step(self, train_batch, batch_idx):
+        """
+        Override `training_step` from PredictionAlgorithm class for CACM-specific training loop.
+
+        """
+
+        self.featurizer = self.model[0]
+        self.classifier = self.model[1]
+
+        minibatches = train_batch
+
+        objective = 0
+        correct, total = 0, 0
+        penalty_causal, penalty_conf, penalty_ind, penalty_sel = 0, 0, 0, 0
+        nmb = len(minibatches)
+
+        features = [self.featurizer(xi) for xi, _, _ in minibatches]
+        classifs = [self.classifier(fi) for fi in features]
+        targets = [yi for _, yi, _ in minibatches]
+
+        for i in range(nmb):
+            objective += F.cross_entropy(classifs[i], targets[i])
+            correct += (torch.argmax(classifs[i], dim=1) == targets[i]).float().sum().item()
+            total += classifs[i].shape[0]
+
+        objective /= nmb
+        loss = objective
+
+        if self.attr_types != []:
+            for attr_type_idx, attr_type in enumerate(self.attr_types):
+                attribute_labels = [
+                    ai for _, _, ai in minibatches
+                ]  # [(batch_size, num_attrs)_1, batch_size, num_attrs)_2, ..., (batch_size, num_attrs)_(num_environments)]
```

Comment:
remove the commented code if not needed

Reply:
Added this to provide greater clarity on the variable shape. Can remove if you feel it's not needed.

Reply:
Okay, it is good then.